Добавляем лазгану возможность менять батарейки

Nachtjager opened the pull request:
Теперь из лазгана можно вынуть батарейку, а потом вставить.

Запил на будущее для всех энергопушек без замозарядки. Если одобрят.

Nachtjager committed:
Добавляем лазгану возможность менять батарейки

Теперь из лазгана можно вынуть батарейку, а потом вставить.

Запил на будущее для всех энергопушек без замозарядки. Если одобрят.

RenardQueenston commented:
Может запилить фичу, что не просто вынимаем батарейку, а откручиваем и прикручиваем её отвёрткой? :thinking:

Nachtjager commented:
В процессе

RenardQueenston commented:
Ну или хотя бы таймер на снятие батарейки и на вставляние.
Типа, чтобы не мгновенно это было.

Nachtjager commented:
Я уже почти закончил с отвёрткой.

RenardQueenston commented:
Умная мысля приходит опосля...
Я бы привязал к скиллу weapon experience скорость изъятия и вставляния (замены, в общем) батарейки.
С отверткой тоже можно.. Но с ней лучше тогда попробовать запилить колхозный, но быстрый, метод изъятия, после которого батарейка становится недееспособной(типа выковыриваешь её). Не уверен в колхозе, поэтому лучше просто прикрутить таймер depending on skill level.

Nachtjager committed:
Update laser.dm

Nachtjager committed:
Теперь для вынимания батарейки требуется отвёртка

Nachtjager committed:
Теперь отвинчивание крышки требует 0.5*на уровнь навыка (unskilled берётся за 1) секунд

Nachtjager committed:
Заканчиваем фичу с батарейкой.

Теперь существует 3 типа (малая, средняя, большая) и 2 тира (обычная, расширеная) батареек для пушек.

Каждой пушке нужен свой размер. Нельзя пихнуть в лазерную пушку среднюю батарейку. Только большую.

Расширенная батарейка будет увеличивать ёмкость оружия в 2 раза что будет двавать в 2 раза больше выстрелов.

Поправлено всё оружие которое есть у людей (и немного у скреллов).

Из пушек используемых боргами и РИГами не могут быть вытащены батарейки.

Так же они не могут быть вытащены из пушек которые не используют чётко заданные батарейки, а используют сгенеренные при спавне.

Время откручивания крышечки отвёрткой убавилось до 0.5 с.
Теперь есть время на извлечение и добавлене батарейки. От 1 (мастер) до 5 (полный 0) секунд

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+    . = ..(user)
+    if (cell_type != null)
+        var/obj/item/weapon/cell/C = cell_type
+        to_chat(user, "This gun use [C.name] as a power source.")
```
```suggestion
        to_chat(user, "This gun uses \a [C.name] as \a power source.")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+
+/obj/item/weapon/gun/energy/proc/unload_battary(mob/user)
+    if(cell_type == null || use_external_power)
+        to_chat(user, "<span class='warning'>You can't pull out battery from [src].</span>")
```
(не уверен, не будет ли дублироваться частица the, лучше проверить)
```suggestion
        to_chat(user, "<span class='warning'>You can't pull \the battery out of \the [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+        to_chat(user, "<span class='warning'>You can't pull out battery from [src].</span>")
+        return
+    if(!hatch)
+        to_chat(user, "<span class='warning'>Open a hatch firts.</span>")
```
```suggestion
        to_chat(user, "<span class='warning'>Open \the hatch first.</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+    if (power_supply)
+        if(do_after(user, 10 * (SKILL_MAX + 1 - user.get_skill_value(SKILL_WEAPONS)) , src))
+            user.put_in_hands(power_supply)
+            user.visible_message("[user] removes [power_supply] from [src].", "<span class='notice'>You remove [power_supply] from [src].</span>")
```
Проверить дублирование the
```suggestion
            user.visible_message("[user] removes \a [power_supply] from \the [src].", "<span class='notice'>You remove \a [power_supply] from \the [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+            power_supply = null
+            on_update_icon()
+    else
+        to_chat(user, "<span class='warning'>[src] is empty.</span>")
```
```suggestion
        to_chat(user, "<span class='warning'>\The [src] is empty.</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+
+/obj/item/weapon/gun/energy/proc/load_battary(var/obj/item/B, mob/user)
+    if(cell_type == null || use_external_power)
+        to_chat(user, "<span class='warning'>You can't insert out battery in [src].</span>")
```
```suggestion
        to_chat(user, "<span class='warning'>You can't insert \a battery into \the [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+    if(istype(B, cell_type))
+        . = TRUE
+        if(!hatch)
+            to_chat(user, "<span class='warning'>Open a hatch firts.</span>")
```
```suggestion
            to_chat(user, "<span class='warning'>Open \the hatch first.</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+            to_chat(user, "<span class='warning'>Open a hatch firts.</span>")
+            return
+        if(power_supply)
+            to_chat(user, "<span class='warning'>[src] already has a battry loaded.</span>")
```
Проверить дублирование the
```suggestion
            to_chat(user, "<span class='warning'>\The [src] already has \a battery loaded.</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+            if(!user.unEquip(B, src))
+                return
+            power_supply = B
+            user.visible_message("[user] inserts [B] into [src].", "<span class='notice'>You insert [B] into [src].</span>")
```
```suggestion
            user.visible_message("[user] inserts \a [B] into \the [src].", "<span class='notice'>You insert \a [B] into \the [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+            playsound(loc, bat_insert_sound, 50, 1)
+            on_update_icon()
+    else if (istype(B, /obj/item/weapon/cell))
+        to_chat(user, "<span class='warning'>You can't use this cell in [src].</span>")
```
```suggestion
        to_chat(user, "<span class='warning'>You can't use this cell in \the [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+    else if (istype(B, /obj/item/weapon/cell))
+        to_chat(user, "<span class='warning'>You can't use this cell in [src].</span>")
+    else
+        to_chat(user, "<span class='warning'>You can't use this in [src]. Stupid man.</span>")
```
Так будет гендерно-нейтрально, к тому же
```suggestion
        to_chat(user, "<span class='warning'>You can't use this in \the [src], stupid.</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+/obj/item/weapon/gun/energy/proc/togle_hatch(mob/user)
+    if(do_after(user, 5, src))
+        if(!hatch)
+            user.visible_message("[user] open hatch on [src].", "<span class='notice'>You open hatch on [src].</span>")
```
```suggestion
            user.visible_message("[user] opens \the hatch on \the [src].", "<span class='notice'>You open \the hatch on \the [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,73 @@
+            user.visible_message("[user] open hatch on [src].", "<span class='notice'>You open hatch on [src].</span>")
+            hatch = TRUE
+        else
+            user.visible_message("[user] close hatch on [src].", "<span class='notice'>You close hatch on [src].</span>")
```
```suggestion
            user.visible_message("[user] closes \the hatch on \the [src].", "<span class='notice'>You close \the hatch on \the [src].</span>")
```

AshtonFox reviewed (changes requested):
Следует исправить грамматику и проверить корректность отображения сообщений в игре.

Nachtjager committed:
Финальные правки

Правки во всех сообщениях.

Добавлено сообщение при эксмайне пушки, которое показывает, какие батарейки оно использует.

Nachtjager committed:
Добовляем звук работы отвёрткой.

Nachtjager committed:
Фикс отвёртки. И её консервирование.

Nachtjager committed:
Отключаем проверку на лючок. Меняем таймер.

Теперь время изъятия батарейки варьиурется от 3.5 до 0.7 секунды. Время уменьшается на 0.7 секунды за уровеь навыка.

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,84 @@
+/obj/item/weapon/gun/energy/examine(mob/user)
+    . = ..(user)
+    if (!cell_type)
+        to_chat(user, "This weapon use custom energy cell as a power source.")
```
```suggestion
        to_chat(user, "This weapon uses custom energy cell as a power source.")
```

AshtonFox committed:
Update frontier/code/modules/projectiles/guns/energy.dm

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,84 @@
+    if (power_supply)
+        if(do_after(user, 7 * (SKILL_MAX + 1 - user.get_skill_value(SKILL_WEAPONS)) , src))
+            user.put_in_hands(power_supply)
+            user.visible_message("[user] removes a [power_supply] from the [src].", "<span class='notice'>You remove a [power_supply] from the [src].</span>")
```
```suggestion
            user.visible_message("[user] removes [power_supply] from [src].", "<span class='notice'>You remove [power_supply] from [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,84 @@
+
+/obj/item/weapon/gun/energy/proc/unload_battary(mob/user)
+    if(cell_type == null || use_external_power)
+        to_chat(user, "<span class='warning'>You can't pull the battery out of the [src].</span>")
```
```suggestion
        to_chat(user, "<span class='warning'>You can't pull the battery out of [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,84 @@
+
+/obj/item/weapon/gun/energy/proc/load_battary(var/obj/item/B, mob/user)
+    if(cell_type == null || use_external_power)
+        to_chat(user, "<span class='warning'>You can't insert a battery into the [src].</span>")
```
```suggestion
        to_chat(user, "<span class='warning'>You can't insert a battery into [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,84 @@
+            to_chat(user, "<span class='warning'>Open the hatch first.</span>")
+            return
+        if(power_supply)
+            to_chat(user, "<span class='warning'>The [src] already has a battery loaded.</span>")
```
```suggestion
            to_chat(user, "<span class='warning'>[src] already has a battery loaded.</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,84 @@
+            if(!user.unEquip(B, src))
+                return
+            power_supply = B
+            user.visible_message("[user] inserts a [B] into the [src].", "<span class='notice'>You insert a [B] into the [src].</span>")
```
```suggestion
            user.visible_message("[user] inserts [B] into [src].", "<span class='notice'>You insert [B] into [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,84 @@
+            playsound(loc, bat_insert_sound, 50, 1)
+            on_update_icon()
+    else if (istype(B, /obj/item/weapon/cell))
+        to_chat(user, "<span class='warning'>You can't use this cell in the [src].</span>")
```
```suggestion
        to_chat(user, "<span class='warning'>You can't use this cell in [src].</span>")
```

AshtonFox commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,84 @@
+    else if (istype(B, /obj/item/weapon/cell))
+        to_chat(user, "<span class='warning'>You can't use this cell in the [src].</span>")
+    else
+        to_chat(user, "<span class='warning'>You can't use this in the [src], stupid.</span>")
```
```suggestion
        to_chat(user, "<span class='warning'>You can't use this in [src], stupid.</span>")
```

AshtonFox reviewed (commented):
Всё-таки частицы не нужны.

AshtonFox committed:
Update frontier/code/modules/projectiles/guns/energy.dm

AshtonFox committed:
Update frontier/code/modules/projectiles/guns/energy.dm

AshtonFox committed:
Update frontier/code/modules/projectiles/guns/energy.dm

AshtonFox committed:
Update frontier/code/modules/projectiles/guns/energy.dm

AshtonFox committed:
Update frontier/code/modules/projectiles/guns/energy.dm

AshtonFox committed:
Update frontier/code/modules/projectiles/guns/energy.dm

AshtonFox committed:
Update frontier/code/modules/projectiles/guns/energy.dm

AshtonFox reviewed (approved):
Yes...

AlexMorgan3817 commented on frontier/code/modules/projectiles/guns/energy.dm:
```diff
@@ -0,0 +1,84 @@
+    if (power_supply)
+        if(do_after(user, 7 * (SKILL_MAX + 1 - user.get_skill_value(SKILL_WEAPONS)) , src))
+            user.put_in_hands(power_supply)
+            user.visible_message("[user] removes [power_supply] from [src].", "<span class='notice'>You remove [power_supply] from [src].</span>")
```
SPAN_NOTICE()

AshtonFox replied:
Not absolutely necessary but you can open a PR changing these if you like to
